BUG: dot -0. semantics

tylerjereddy opened the pull request:
* Fixes #21342

* add a regression test and fix for the above issue, which seems to boil down to `memset` not being a great fit for negative zero floating point semantics because it zeros out using unsigned char bit patterns

* the testsuite was happy about this solution, but I'll add one or two comments inline re: residual concerns

* @stefanv helped me with a bit of debugging here as well, during the SciPy sprint

tylerjereddy committed:
BUG: dot -0. semantics

* Fixes #21342

* add a regression test and fix for the above issue,
which seems to boil down to `memset` not being a great
fit for negative zero floating point semantics because
it zeros out using unsigned char bit patterns

* the testsuite was happy about this solution, but I'll
add one or two comments inline re: residual concerns

tylerjereddy commented on numpy/core/src/common/cblasfuncs.c:
```diff
@@ -387,6 +387,10 @@ cblas_matrixproduct(int typenum, PyArrayObject *ap1, PyArrayObject *ap2,
             else if (ap1shape != _matrix) {
+                for (size_t i = 0; i < PyArray_SIZE(out_buf); i += 1) {
+                    ((double *)PyArray_DATA(out_buf))[i] = -0.0;
+                }
```
Two thoughts I had here:

1) I'm a bit surprised the testsuite is happy with this (locally, at least)--I would have thought it may be risky not to use one of the C-API iterator approaches because of strides and so on?
2) If we're going to commit to this, we may want to replace the original `memset` entirely and hoist this farther up in the control flow--I went a conservative to the specific code path in question, at least initially.

seberg replied:
`cblas_matrixproduct` should only be reached with contiguous arrays, I think, so that is fine.

tylerjereddy commented:
Interesting, I see a MacOS ARM failure, which is the platform it passes on for me locally.

tylerjereddy committed:
BUG: PR 24197 revisions

* the CI failures were caused by cases where OpenBLAS
was absent and we traversed the `dot` codepath that used one
of our templated C inner loops, which I've now fixed
to initialize the sum to `-0.0`

* expand test to include some striding variety, 0-D, and
size 0 arrays; for the dot product of two empty arrays
I didn't have a compelling reason to require anything other
than "positive" zero (I'm not aware of any references on this
one, and we have no values to round up or down from)

tylerjereddy commented:
Some attempted improvements today summarized below, but the new test cases don't all pass yet.

```
    * the CI failures were caused by cases where OpenBLAS
    was absent and we traversed the `dot` codepath that used one
    of our templated C inner loops, which I've now fixed
    to initialize the sum to `-0.0`
    
    * expand test to include some striding variety, 0-D, and
    size 0 arrays; for the dot product of two empty arrays
    I didn't have a compelling reason to require anything other
    than "positive" zero (I'm not aware of any references on this
    one, and we have no values to round up or down from)
```

tylerjereddy committed:
BUG: PR 24197 revisions

* fix up 2-D * scalar signed zero codepath with
double type

* fix up another `dot` inner loop, this time with
CBLAS not getting the correct signed zero semantics

* linter cleanups

tylerjereddy commented:
Ok, this is passing the extended tests now. Some more notes:

- it seems I had to shim around OpenBLAS signed zero semantics for one or two corner cases--eventually upstreaming those changes might be preferable to i.e., adding logic to inner loops where the performance tradeoff for getting the formally-correct sign for zero might be questionable
- I used `npy_signbit` in a few places based on `git grep`--I was aware of some intention to get rid of `npy_math` and leverage C99 fully at some point, but I used the current codebase as a guide at this point
- if there are specific requests for more test cases to make the solution more general, I'm happy to add/address those

seberg commented on numpy/core/src/multiarray/arraytypes.c.src:
```diff
@@ -3653,14 +3653,21 @@ NPY_NO_EXPORT void
+                ((@type@ *)ip1)[0] == 0 &&
+                !npy_signbit(((@type@ *)ip2)[0])) {
+                result = -0.0;
+            }
```
This (and the earlier thing the same way probably), doesn't look right.  `np.dot([-0., 0.], [0., 0.])` should return a positive 0 and I don't think it will here.

I was happy with the idea (and agree that it seems more correct) when I though this is just about initialization inputs with `-0.0` (unless empty).
And initialization is about where I would want to draw the line of contortions to do to get this right.

But...  I can't find a `res = alpha + x.T @ y` in level 1 BLAS.  So, this isn't possible beyond (potentially, very unlikely) inspecting all input values and redoing most of the calculation.  (Only when the result is `0.0`).  To me, that seems unacceptable considering the magnitude of the issue ;).

So, I am tempted to close the issue with:  Pick it up with BLAS implementors, the best I can offer is a "meh, NumPy dev's somewhat agree that -0.0 should be returned in this cases considering IEEE rules".
(Next thing, do CPU dot intrinsics actually get it right?  My guess is yes, but that is probably where BLAS implementors would draw the line.)

seberg replied:
Sorry, the whole comment refers only the 1-D dot product, I guess.  2-D dot product *can* initialize the output correctly, and I am happy to pursue that a bit more.

(Maybe there is a way to abuse level 2-blas without any slow-down to get this, maybe that would be a point...)
